hatch_pip_compile/plugin.py

```diff
@@ -59,6 +59,17 @@ def __init__(self, *args, **kwargs):
             env_name=self.name,
             project_name=self.metadata.name,
         )
+        install_method = self.config.get("pip-compile-installer", "pip")
+        if install_method == "pip":
+            self.__class__ = PipCompileEnvironmentWithPipInstall
+        elif install_method == "pip-sync":
+            self.__class__ = PipCompileEnvironmentWithPipSync
```

Comment:
I like the idea here with two subclasses that know how to do the right thing. My only hesitation about the approach is I've never overridden `__class__` at `__init__`. It seems similar to how I've seen the `__new__` method be used though instead of `__init__`.

Reply:
Yes. I actually have never done or heard of this 😂 Well I have replaced `__class__` but not from inside the object itself.

`__new__` is definitely the way to go if it works. The big difficulty here is that in that case you don't have `self.config` yet so maybe this parameter would need to be scooped out of the passed config in its `, *args, **kwargs)`.  At which point I figured *that* became messier than this

Reply:
The other option I was considering is to provide 2 separate plugins in 1 package but that can be confusing

Reply:
I'm sure you can see by the `lock.PipCompileLock` class that I'm a big fan of dependency injection. That's how I was thinking about this when I saw the use of `__class__`.

```python
def __init__(*args, **kwargs) -> None:
    super().__init__(*args, **kwargs)
    self.installer = PipCompileEnvironmentWithPipInstall(config=self.config)
    
def install_project(self) -> None:
    self.installer.install_project()
```

I'm sure this would require a bunch of passed attributes though and doesn't work as nicely as as a as a straight up subclass does.

Reply:
What would you think about something like this: https://github.com/juftin/hatch-pip-compile/commit/bfb243442809f0a8d00f1b94714ea84f7ac4591e

My idea here is that the `installer` attribute gets direct access to the environment itself. Same general idea of making the `PipCompileEnvironment` class impartial to how its dependencies get installed but without overriding the insternal `__class__` attribute

Reply:
I was initially skeptical and wrote this

> The vast majority of the benefit I got from subclassing is being able to choose only 1 method to override in 1 case and many methods to override in the other case.

But actually it seems because we happen to already override `install_project_dev_mode` anyway even for the simple case, it all works out. I think it looks good, will look in a bit more detail later.

Reply:
This is fun!
I had some thinking about properly keeping each file's concerns to itself - most importantly that the "installers" file should never mention pip-compile, for example.

Here are my suggestions, please check them commit by commit:
https://github.com/oprypin/hatch-pip-compile/compare/b6b93de15afb06ba1d15f7fdb918078ccbe7ac11...pip-split2

Or the whole branch
https://github.com/oprypin/hatch-pip-compile/compare/pip-split2

I still didn't test this too carefully though.

And how are we going to merge these branches on branches? :D  Anyway we should probably put this into a pull request of yours, and then I guess squash, with me as a sub-author

Reply:
Oh I like these changes - that's a good API design. I will run some testing and confirm a few things. I'm excited for for some automated tests soon (starting with #18).

As far as merging goes I was going to just open a PR from [pip-split2 to main](https://github.com/juftin/hatch-pip-compile/compare/main...oprypin:hatch-pip-compile:pip-split2) and squash merge it. Happy to do that differently if you prefer. 

Similarly, I'm super proud and excited about the work from this feature. I really feel like this thing is ready for prime-time now that we support these different install options. Thank you so much for all the thought and contributions on everything.

Reply:
I was considering replacing `python` with this:

```python
executable = (
    pathlib.Path(self.virtual_env.executables_directory)
    / self.virtual_env.python_info.executable
)
``` 

But I'm wondering if that might cause an issue with Windows.... I guess I'll test in CI/CD

Reply:
Now that the `environment.safe_activation()` issue is resolved I'm seeing another error surfaced in this commit: https://github.com/juftin/hatch-pip-compile/commit/e5d9c50aa27d828e8d2578d9d461e16690fbfc58

<details><summary>diff</summary>
<p>

```diff
diff --git a/hatch_pip_compile/plugin.py b/hatch_pip_compile/plugin.py
index 9fb8282..17d1cf5 100644
--- a/hatch_pip_compile/plugin.py
+++ b/hatch_pip_compile/plugin.py
@@ -107,12 +107,13 @@ class PipCompileEnvironment(VirtualEnvironment):
         Run pip-compile if necessary
         """
         if not self.lockfile_up_to_date:
-            self.install_pip_tools()
-            if self.piptools_lock_file.exists():
-                _ = self.piptools_lock.compare_python_versions(
-                    verbose=self.config.get("pip-compile-verbose", None)
-                )
-            self.pip_compile_cli()
+            with self.safe_activation():
+                self.install_pip_tools()
+                if self.piptools_lock_file.exists():
+                    _ = self.piptools_lock.compare_python_versions(
+                        verbose=self.config.get("pip-compile-verbose", None)
+                    )
+                self.pip_compile_cli()

     def pip_compile_cli(self) -> None:
         """
diff --git a/pyproject.toml b/pyproject.toml
index 12aecd3..ccc1e46 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -67,6 +67,8 @@ python = ["3.8", "3.9", "3.10", "3.11", "3.12"]

 [tool.hatch.envs.default]
 pip-compile-constraint = "default"
+pip-compile-verbose = true
+pip-compile-installer = "pip-sync"
 post-install-commands = [
   "- pre-commit install"
 ]
```

</p>
</details> 

## Issue

```
Usage: python -m piptools sync [OPTIONS] [SRC_FILES]...
Try 'python -m piptools sync -h' for help.

Error: Invalid value for '[SRC_FILES]...': Path '/Users/justinflannery/git/juftin/hatch-pip-compile/requirements/requirements-docs.txt' does not exist.
```

## Steps to Reproduce

set `pip-compile-installer` to `pip-sync`

```
rm -rf .venv/ requirements/ requirements.txt
hatch run docs:serve
```

The default lockfile is created and installed but `pip-sync` is called on the docs lockfile before it can be created

Reply:
Okay 😅 see https://github.com/juftin/hatch-pip-compile/pull/35

There are two small changes that resolve the issues I mentioned in the thread